modules/network_synchronizer/networked_controller.h

```diff
@@ -0,0 +1,389 @@
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+/**
+	@author AndreaCatania
+*/
+
+#include "scene/main/node.h"
+
+#include "core/math/transform.h"
+#include "core/node_path.h"
+#include "input_buffer.h"
+#include <deque>
+#include <vector>
+
+#ifndef NETWORKED_CONTROLLER_H
+#define NETWORKED_CONTROLLER_H
+
+struct Controller;
+class SceneRewinder;
+
+/// The `NetworkedController` is responsible to sync the `Player` inputs between
+/// the peers. This allows to control a character, or an object with high precision
+/// and replicates that movement on all connected peers.
+///
+/// The `NetworkedController` will sync inputs, based on those will perform
+/// operations.
+/// The result of these operations, are guaranteed to be the same accross the
+/// peers, if we stay under the assumption that the initial state is the same.
+///
+/// Is possible to use the `SceneRewinder` to keep the state in sync with the
+/// peers.
+///
+// # Implementation details
+//
+// The `NetworkedController` perform different operations depending where it's
+// instantiated.
+// The most important part is inside the `PlayerController`, `ServerController`,
+// `DollController`, `NoNetController`.
+class NetworkedController : public Node {
+	GDCLASS(NetworkedController, Node);
+
+public:
+	enum InputCompressionLevel {
+		INPUT_COMPRESSION_LEVEL_0 = InputsBuffer::COMPRESSION_LEVEL_0,
+		INPUT_COMPRESSION_LEVEL_1 = InputsBuffer::COMPRESSION_LEVEL_1,
+		INPUT_COMPRESSION_LEVEL_2 = InputsBuffer::COMPRESSION_LEVEL_2,
+		INPUT_COMPRESSION_LEVEL_3 = InputsBuffer::COMPRESSION_LEVEL_3
+	};
+
+private:
+	/// The input storage size is used to cap the amount of inputs collected by
+	/// the `Master`.
+	///
+	/// The server sends a message, to all the connected peers, notifing its
+	/// status at a fixed interval.
+	/// The peers, after receiving this update, removes all the old inputs until
+	/// that moment.
+	///
+	/// If the `input_storage_size` is too small, the clients will collect inputs
+	/// intermittently, but on the other side, a too large value may introduce
+	/// virtual delay.
+	///
+	/// With 60 iteration per seconds a good value is `300`, but is adviced to
+	/// perform some tests until you find a better suitable value for your needs.
+	int player_input_storage_size;
+
+	/// Amount of time an inputs is re-sent to each peer.
+	/// Resend inputs is necessary because the packets may be lost since they
+	/// are sent in an unreliable way.
```

Comment:
```suggestion
	/// Amount of time inputs are re-sent to each peer.
	/// Resending inputs is necessary because the packets may be lost as they
	/// are sent in an unreliable way.
```

Reply:
Done